lib/ProductOpener/Store.pm

```diff
@@ -255,50 +277,269 @@ sub retrieve ($file) {
+	if (-e $file_path) {
+		my $json;
+		eval {
+			open(my $IN, "<", $file_path) or die("Can't open $file_path");
+			flock($IN, LOCK_SH);
+			local $/;    #Enable 'slurp' mode
+			$json = <$IN>;
+			# Release the lock. Some docs say this isn't needed but tests show otherwise
+			flock($IN, LOCK_UN);
+			close($IN);
+		} or do {
+			$log->error("retrieve_object", {path => $path, error => $@}) if $log->is_error();
+		};
+		return $json;
+	}
+	# Fallback to old method
+	return $json_for_objects->encode(retrieve($path . '.sto'));
+}
 
-	return $return;
+=head2 object_exists($path)
+
+Indicates whether an object (STO or JSON) exists at the specified path
+
+=cut
+
+sub object_exists($path) {
+	return (-e "$path.json" or -e "$path.sto");
+}
+
+=head2 object_path_exists($path)
+
+Indicates whether an directory exists at the specified path
+
+=cut
+
+sub object_path_exists($path) {
+	return (-d $path);
+}
+
+=head2 move_object($old_path, $new_path)
+
+Moves a single object or all objects in the path
+
+=cut
+
+sub move_object($old_path, $new_path) {
+	# File::Copy move() is intended to move files, not
+	# directories. It does work on directories if the
+	# source and target are on the same file system
+	# (in which case the directory is just renamed),
+	# but fails otherwise.
+	# An alternative is to use File::Copy::Recursive
+	# but then it will do a copy even if it is the same
+	# file system...
+	# Another option is to call the system mv command.
+	$log->debug("moving object data", {source => $old_path, destination => $new_path})
+		if $log->is_debug();
+
+	if (-d $old_path) {
+		# Moving a while directory
+		ensure_dir_created_or_die($new_path);
+		#11872 TOD Should probably die here
+		dirmove($old_path, $new_path)
+			or $log->error("could not move objects", {source => $old_path, destination => $new_path, error => $!});
+
+	}
+	else {
+		# Moving a single file
+		ensure_dir_created_or_die(dirname($new_path));
+		if (-e "$old_path.sto") {
+			move("$old_path.sto", "$new_path.sto")
+				or die("could not move sto file from $old_path to $new_path, error: $!");
+		}
+		else {
+			move("$old_path.json", "$new_path.json")
+				or die("could not move json file from $old_path to $new_path, error: $!");
+		}
+	}
+
+	return;
 }
 
-=head2  sto_iter($initial_path, $pattern=qr/\.sto$/i)
+=head2 link_object($name, $link)
 
-iterate all the files corresponding to $pattern starting from $initial_path
+Makes the $link point to the data in the specified relative $path.
+If the object at the $path is an sto file then an STO symbolic link will be created
+
+=cut
 
-use it as an iterator:
-my $iter = sto_iter(".");
-while (my $path = $iter->()) {
-	# do stuff
+sub link_object($name, $link) {
+	# If target is a sto file then keep the link as a sto file too. Note we use relative paths for the target file
+	#11901: Remove $serialize_to_json test once production is migrated
+	if (not $serialize_to_json or -e dirname($link) . '/' . $name . '.sto') {
```

Comment:
Should we definitely not add symlinks for existing STO files after migration begins?  I don't yet totally understand what the symlinks are for -- so maybe this is correct, but I figured I'd ask.